src/libraries/System.Text.Json/tests/Common/JsonSchemaExporterTests.TestTypes.cs

```diff
@@ -0,0 +1,1320 @@
+                  "properties": {
+                    "Y": { "type": "string" },
+                    "Z": { "type": "integer" },
+                    "X": { "type": "string" }
+                  },
+                  "required": [ "Y", "Z", "X" ]
+                }
+                """);
+
+            yield return new TestData<PocoWithIgnoredMembers>(new() { X = 1, Y = 2 }, ExpectedJsonSchema: """{"type":["object","null"],"properties":{"X":{"type":"integer"}}}""");
+            yield return new TestData<PocoWithCustomNaming>(
+                Value: new() { IntegerProperty = 1, StringProperty = "str" },
+                ExpectedJsonSchema: """
+                {
+                  "type": ["object","null"],
+                  "properties": {
+                    "int": { "type": "integer" },
+                    "str": { "type": ["string", "null"] }
+                  }
+                }
+                """);
+
+            yield return new TestData<PocoWithCustomNumberHandling>(
+                Value: new() { X = 1 },
+                ExpectedJsonSchema: """
+                {
+                    "type": ["object","null"],
+                    "properties": { "X": { "type": ["string","integer"] } }
+                }
+                """);
+
+            yield return new TestData<PocoWithCustomNumberHandlingOnProperties>(
+                Value: new() { X = 1, Y = 2, Z = 3 },
+                AdditionalValues: [
+                    new() { X = 1, Y = double.NaN, Z = 3 },
+                    new() { X = 1, Y = double.PositiveInfinity, Z = 3 },
+                    new() { X = 1, Y = double.NegativeInfinity, Z = 3 },
+                ],
+                ExpectedJsonSchema: """
+                {
+                  "type": ["object","null"],
+                  "properties": {
+                    "X": { "type": ["string", "integer"] },
+                    "Y": {
+                      "anyOf": [
+                        { "type": "number" },
+                        { "enum": ["NaN", "Infinity", "-Infinity"]}
+                      ]
+                    },
+                    "Z": { "type": ["string", "integer"] },
+                    "W" : { "type": "number" }
+                  }
+                }
+                """);
+
+            yield return new TestData<PocoWithRecursiveMembers>(
+                Value: new() { Value = 1, Next = new() { Value = 2, Next = new() { Value = 3 } } },
+                AdditionalValues: [new() { Value = 1, Next = null }],
+                ExpectedJsonSchema: """
+                {
+                    "type": ["object","null"],
+                    "properties": {
+                        "Value": { "type": "integer" },
+                        "Next": {
+                            "type": ["object", "null"],
+                            "properties": {
+                                "Value": { "type": "integer" },
+                                "Next": { "$ref": "#/properties/Next" }
```

Comment:
I've updated the implementation so that `$ref`'s are only emitted for recursive types. My understanding is that a `$ref` is the only way such schemas can expressed (independent of whether they have been factored into a `$defs` section).

Reply:
> My understanding is that a $ref is the only way such schemas can expressed

Correct.

> ... independent of whether they have been factored into a $defs section

_Technically_ correct, but again, not recommended best practice.

My concern is that other developers, seeing that .Net is doing this, will interpret it as a recommended approach (at least in .Net spheres) regardless of the fact that _JSON Schema_ recommends against it.

I strongly urge you to extract reference targets (that aren't the root schema) to `$defs`.

Alternatively, the `$anchor` or `$id` approach _as a default_ is acceptable.

Reply:
Understood, at the same time I wanted to avoid shipping a baked in naming convention (since invariably that would involve either generating UUID's for every node in the type graph or leaking fully qualified type names to the schema). Using JSON pointer focuses on the structural aspect of the generated document, and users should be able to convert that to whatever naming convention they desire using the callback.

My main concern with the current approach is that JSON pointer currently only supports absolute paths, which causes it to break if composed inside other documents, putting the responsibility on users to fix it up. Because `$ref` is only generated for recursive types, a relative pointer would simply need to specify the depth of the recursive occurrence.